runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillTimerInternals.java

```diff
@@ -94,8 +98,25 @@ public void setTimer(TimerData timerKey) {
 
   @Override
   public void setTimer(
-      StateNamespace namespace, String timerId, Instant timestamp, TimeDomain timeDomain) {
-    timers.put(timerId, namespace, TimerData.of(timerId, namespace, timestamp, timeDomain));
+      StateNamespace namespace,
+      String timerId,
+      Instant timestamp,
+      Instant outputTimestamp,
+      TimeDomain timeDomain) {
+
+    TimerData timerData = TimerData.of(timerId, namespace, timestamp, outputTimestamp, timeDomain);
+
+    if (WindmillNamespacePrefix.USER_NAMESPACE_PREFIX.equals(prefix)) {
+      outputBuilder
+          .addWatermarkHoldsBuilder()
+          .setTag(timerHoldTag(prefix, timerData))
+          .setStateFamily(stateFamily)
+          .setReset(true)
+          .addTimestamps(
+              WindmillTimeUtils.harnessToWindmillTimestamp(timerData.getOutputTimestamp()));
+    }
```

Comment:
I don't think it's correct to modify the outputBuilder here. I think that this logic belongs in the persistTo method instead

Reply:
@reuvenlax  we need more description on how we can add watermark hold here. We look into persistTo method and was not able to find any clue.

Reply:
persistTo already has code to set a watermark hold. If you change the hold from timerData.getTimestamp to timerData.getOutputTimestamp, does that not work to set the hold?

Reply:
I think it should work because similar work was done in StatefulParDoEvaluatorFactory where outputTimestamp is used to set required holds.